libvast/vast/plugin.hpp

```diff
@@ -0,0 +1,177 @@
+  using analyzer_actor
+    = caf::typed_actor<caf::reacts_to<caf::stream<table_slice>>>;
+
+  /// Creates an actor that hooks into the input table slice stream.
+  /// @param sys The actor system context to spawn the actor in.
+  /// @returns The actor handle to the analyzer.
+  virtual analyzer_actor make_analyzer(caf::actor_system& sys) const = 0;
+};
+
+// -- command plugin -----------------------------------------------------------
+
+/// A base class for plugins that add commands.
+/// @relates plugin
+class command_plugin : public virtual plugin {
+public:
+  /// Creates additional commands.
+  virtual std::pair<std::unique_ptr<command>, command::factory>
+  make_command() const = 0;
+};
+
+// -- plugin_ptr ---------------------------------------------------------------
+
+/// An owned plugin and dynamically loaded plugin.
+/// @relates plugin
+class plugin_ptr final {
+public:
+  /// Load a plugin from the specified library filename.
+  /// @param filename The filename that's passed to 'dlopen'.
+  explicit plugin_ptr(const char* filename) noexcept;
+
+  /// Unload a plugin and its required resources.
+  ~plugin_ptr() noexcept;
+
+  /// Forbid copying of plugins.
+  plugin_ptr(const plugin_ptr&) = delete;
+  plugin_ptr& operator=(const plugin_ptr&) = delete;
+
+  /// Move-construction and move-assignment.
+  plugin_ptr(plugin_ptr&& other) noexcept;
+  plugin_ptr& operator=(plugin_ptr&& rhs) noexcept;
+
+  /// Pointer facade.
+  explicit operator bool() noexcept;
+  const plugin* operator->() const noexcept;
+  plugin* operator->() noexcept;
+  const plugin& operator*() const noexcept;
+  plugin& operator&() noexcept;
+
+  /// Upcast a plugin to a more specific plugin type.
+  /// @tparam Plugin The specific plugin type to try to upcast to.
+  /// @returns A pointer to the upcasted plugin, or 'nullptr' on failure.
+  template <class Plugin>
+  const Plugin* as() const {
+    static_assert(std::is_base_of_v<plugin, Plugin>, "'Plugin' must be derived "
+                                                     "from 'vast::plugin'");
+    return dynamic_cast<const Plugin*>(instance_);
+  }
+
+  /// Upcast a plugin to a more specific plugin type.
+  /// @tparam Plugin The specific plugin type to try to upcast to.
+  /// @returns A pointer to the upcasted plugin, or 'nullptr' on failure.
+  template <class Plugin>
+  Plugin* as() {
+    static_assert(std::is_base_of_v<plugin, Plugin>, "'Plugin' must be derived "
+                                                     "from 'vast::plugin'");
+    return dynamic_cast<Plugin*>(instance_);
+  }
+
+private:
+  /// Implementation details.
+  void* library_ = {};
+  plugin* instance_ = {};
+  void (*deleter_)(plugin*) = {};
+};
+
+} // namespace vast
+
+// -- helper macros ------------------------------------------------------------
+
+#define VAST_REGISTER_PLUGIN(name, major, minor, tweak, patch)                 \
+  extern "C" ::vast::plugin* plugin_create() {                                 \
```

Comment:
Would it make sense to call these `vast_plugin_create()`, etc.?

Reply:
Yes!